### PR TITLE
8284493: Improve computeNextExponential tail performance and accuracy

### PR DESCRIPTION
**Pull request by Pr0methean**
This PR improves both the worst-case performance of `nextExponential` and `nextGaussian` and the distribution of output at the tails. It fixes the following imperfections:

* Repeatedly adding DoubleZigguratTables.exponentialX0 to extra causes a rounding error to accumulate at the tail of the distribution (probably starting around `2*exponentialX0 == 0x1.e46eff20739afp3 ~ 15.1`); this PR fixes that by tracking the multiple of exponentialX0 as a long. (This distortion is worst when `x > 0x1.0p56` since in that case, a rounding error means `extra + x == extra`.
* Reduces several equations using `Math.fma`. (This will almost certainly improve performance, and may or may not improve output distribution.)
* Uses the newly-extracted `computeWinsorizedNextExponential` function to prevent `nextGaussian` from going into the `nextExponential` tail twice.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8284493](https://bugs.openjdk.org/browse/JDK-8284493): Improve computeNextExponential tail performance and accuracy (**Enhancement** - P4)

### Reviewers
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/8131/head:pull/8131` \
`$ git checkout pull/8131`

Update a local copy of the PR: \
`$ git checkout pull/8131` \
`$ git pull https://git.openjdk.org/jdk.git pull/8131/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8131`

View PR using the GUI difftool: \
`$ git pr show -t 8131`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8131.diff">https://git.openjdk.org/jdk/pull/8131.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/8131#issuecomment-1122475537)

### CODE REVIEW

**Comment by navyxliu** [comment]
/covered
@Pr0methean is an employee of Amazon.

**Comment by navyxliu** [comment]
hi, @JesperIRL, 
Could you help us on OCA coverage?

**Comment by Pr0methean** [comment]
Should we extract a computeWinsorizedNextExponential so that line 1218 can realistically be covered in a unit test?

**Comment by simonis** [comment]
/covered

@robilad, can you please verify Chris Hennick's (@Pr0methean) OCA status?
He's an Amazon employee and covered by Amazons OCA.

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java
Comment (turbanoff) [inline src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1411]: Code is not compilable

Comment (rgiulietti) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1411]: @Pr0methean As @turbanoff observes, I think there's a closing parentheses too much on L.1411 and one on L.1421.
Which compiler are you using ;-)  ?

Comment (Pr0methean) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java]: Should be fixed as of fa340fb47b1169805f21fb71cce0da3fc175d427. I haven't yet found a way for IntelliJ IDEA to use the project as its own JDK, so I've sanity-checked the syntax using the Rainbow Brackets plugin.

Comment (Pr0methean) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java]: Update: successfully built `images` and `run-test-tier1` on my AL2 machine.

**Comment by Pr0methean** [comment]
In addition to the changes discussed heretofore, I've also changed line 1382 to eliminate unneeded tail exploration; this should make `nextGaussian` faster at p99 and up.

**Comment by Pr0methean** [comment]
Keep open; I'll submit some benchmarks when I have a chance.

**Comment by Pr0methean** [comment]
Keep open

**Comment by Pr0methean** [comment]
Keep open

**Comment by Pr0methean** [comment]
Update: just waiting for approval from work to obtain c6a.metal, c6i.metal and c6g.metal instances for long enough to run the benchmarks before and after.

**Comment by Pr0methean** [comment]
Keep open

**Comment by Pr0methean** [comment]
Keep open. Testing is in progress right now.

**Comment by Pr0methean** [comment]
JMH benchmark results (`make test TEST="micro:java.util.random"`) with `fixedSeed = true` on EC2 are below. Average time improves only for nextExponential() and only on non-Intel CPUs, but p100 improved more consistently (7 of the 12 tests improved by at least 1000ns, only 2 deteriorated by that amount.)
# c6i.metal
L64X128MixRandom.nextExponential() average ns/op: 26.984 ± 0.015 before, 34.914 ± 0.021 after
L64X128MixRandom.nextExponential() p100 ns/op: 8224 before, 5624 after
L64X1024MixRandom.nextExponential() average ns/op: 30.026 ± 0.020 before, 36.730 ± 0.024 after
L64X1024MixRandom.nextExponential() p100 ns/op: 8736 before, 7464 after
L64X128MixRandom.nextGaussian() average ns/op: 29.995 ± 0.017 before, 30.016 ± 0.019 after
L64X128MixRandom.nextGaussian() p100 ns/op: 5392 before, 3024 after
L64X1024MixRandom.nextGaussian() average ns/op: 31.394 ± 0.020 before, 31.660 ± 0.021 after
L64X1024MixRandom.nextGaussian() p100 ns/op: 5984 before, 6072 after
# c7g.16xlarge
L64X128MixRandom.nextExponential() average ns/op: 41.006 ± 0.056 before, 40.288 ± 0.055 after
L64X128MixRandom.nextExponential() p100 ns/op: 6384 before, 6416 after
L64X1024MixRandom.nextExponential() average ns/op: 45.218 ± 0.071 before, 40.549 ± 0.062 after
L64X1024MixRandom.nextExponential() p100 ns/op: 560 before, 3760 after
L64X128MixRandom.nextGaussian() average ns/op: 39.643 ± 0.052 before, 40.473 ± 0.070 after
L64X128MixRandom.nextGaussian() p100 ns/op: 7728 before, 1264 after
L64X1024MixRandom.nextGaussian() average ns/op: 40.393 ± 0.061 before, 40.274 ± 0.061 after
L64X1024MixRandom.nextGaussian() p100 ns/op: 9184 before, 1280 after
# m6a.metal
L64X128MixRandom.nextExponential() average ns/op: 31.241 ± 0.039 before, 29.004 ± 0.018 after
L64X128MixRandom.nextExponential() p100 ns/op: 4224 before, 7224 after
L64X1024MixRandom.nextExponential() average ns/op: 31.903 ± 0.017 before, 31.146 | ± | 0.021 | ns/op |
L64X1024MixRandom.nextExponential() p100 ns/op: 9744 before, 8688 after
L64X128MixRandom.nextGaussian() average ns/op: 29.164 ± 0.017 before, 29.073 ± 0.021 after
L64X128MixRandom.nextGaussian() p100 ns/op: 5920 before, 1504 after
L64X1024MixRandom.nextGaussian() average ns/op: 29.503 ± 0.022 before, 29.639 ± 0.018 after
L64X1024MixRandom.nextGaussian() p100 ns/op: 4256 before, 4288 after

**Comment by navyxliu** [comment]
hi, @Pr0methean 

I notice that  there is a big regression in L64X1024MixRandom.nextExponential p100 on [c7g.16xlarge](https://aws.amazon.com/ec2/instance-types/c7g/)
```
L64X1024MixRandom.nextExponential() p100 ns/op: 560 before, 3760 after
```

but the average ns/op improves. Can you explain the discrepancy? 

btw: I updated the JBS issue to ensure the subject is aligned with your PR. 

Thanks,
--lx

**Comment by Pr0methean** [comment]
It's probably just a statistical fluke involving a very large return value, but to know for sure I'd need a way to gather statistics about the return values and how they correlate with running-time outliers.

**Comment by navyxliu** [comment]
> It's probably just a statistical fluke involving a very large return value, but to know for sure I'd need a way to gather statistics about the return values and how they correlate with running-time outliers.

Yeah, it's likely that p100 data are outliers. I think average and p50 data make more sense. 
I am not a reviewer and lack domain-specific knowledge in the algorithm. I check your code and it seems reasonable. We need reviewers to look into it.

**Comment by Pr0methean** [comment]
@navyxliu p50 isn't meaningful, because this change doesn't make a difference in the ~98% of cases where only the lookup table is used.

**Comment by Pr0methean** [comment]
Keep open.

**Comment by Pr0methean** [comment]
@JimLaskey May I please have a review on this?

**Comment by ysramakrishna** [comment]
cc @jddarcy ... as another potential reviewer. Thanks!

**Comment by Pr0methean** [comment]
Keep open @JimLaskey @jddarcy

**Comment by Pr0methean** [comment]
Keep open.

@JimLaskey @jddarcy May I please have at least a timeline on when to expect a review?

**Comment by Pr0methean** [comment]
@eirbjo Done.

**Comment by Pr0methean** [comment]
@JimLaskey @jddarcy Keep open

**Comment by GuySteele** [comment]
I have now reviewed all the changes in file .../jdk/internal/util/random/RandomSupport.java . It appears that the method formerly called computeWinsorizedNextExponential has been renamed to computeNextExponentialSoftCapped, and that is a good change. All the changes to the executable code appear to be sound (and, yes, I believe I still remember how all this code is supposed to work—even while admitting that I wrote it some years ago and that a couple of embarrassing mistakes were later discovered and repaired, thank goodness). I looked especially carefully at the change of representation for the variable "extra" from double to long and the new uses of fma. I went back and looked at email exchanges I had with Joe Darcy on May 10, 2022, and I see that many of the concerns I expressed at that time have been addressed. 

I do suggest that for extra clarity, the declaration and computation of maxExtraMinus1 at lines 1206--1212 be moved down below to after line 1222, just below the comment that begins "We didn't use the upper part of U1 after all". It may be that good optimizing Java compilers perform this code motion anyway (the code in lines 1213–1221 does not refer to maxExtraMinus1), but it would help a human reader to understand that this code is not part of and does not need to be part of the fast path through the code. Moreover, though a compiler cannot tell that it's okay to move the comparison of maxValue to 0.0 (and the possible forced return of 0.0) off the fast path, I argue that it is indeed okay to do so, because it is always permitted to return a value larger than maxValue, and the fast path does always return a nonnegative value. So I actually argue to move three more lines (in all, lines 1203–1212) to after line 1222.

In all other respects, I recommend that this set of changes be adopted exactly as is.

If making the one change I suggested above might cause adoption to slip from JDK21 to JDK22 (perhaps because of a need for retesting), then I would suggest adopting the code exactly as is and then scheduling the suggested change for JDK22, because the suggested change improves clarity and code speed but should not change the advertised functionality at all.

**Comment by Pr0methean** [comment]
Moved the code after the fast-path return as requested.

**Comment by GuySteele** [comment]
Latest change looks good. I noted the improvement to the comment beginning "We didn't use the upper part of U1 after all".

**Comment by Pr0methean** [comment]
@turbanoff @rgiulietti  Keep open

**Comment by Pr0methean** [comment]
@JimLaskey @jddarcy @turbanoff @rgiulietti Could this please be merged?

**Review by jddarcy — APPROVED** [review]
As Guy is not a capital-R reviewer for this project, marking the PR as Approved on his behalf.

**Comment by jddarcy** [comment]
@Pr0methean , please update your branch with the current contents of the master before issuing the integrate command. Assuming that succeeds as expected, myself or someone else can sponsor your changeset as described in the comment from the Skara bots.

**Comment by Pr0methean** [comment]
@jddarcy Done.

**Pull request closed without merging**
